src/main/java/spoon/support/reflect/declaration/CtMethodImpl.java

```diff
@@ -217,6 +219,35 @@ public boolean isShadow() {
 		return (CtMethod<T>) super.clone();
 	}
 
+	@Override
+	public Collection<CtMethod<?>> getTopDefinitions() {
+		List<CtMethod<?>> s = new ArrayList<>();
+
+		// first collect potential declarations of this method in the type hierarchy
+		ClassTypingContext context = new ClassTypingContext(this.getDeclaringType());
+		for (Object m : getDeclaringType().map(new AllTypeMembersFunction(CtMethod.class)).list()) {
+			if (m != this && context.isOverriding(this, (CtMethod<?>) m)) {
+				s.add((CtMethod) m);
+			}
+		}
+
+		// now removing the intermediate methods for which there exists a definition upper in the hierarchy
+		List<CtMethod<?>> finalMeths = new ArrayList<>(s);
+		for (CtMethod m1 : s) {
+			ClassTypingContext context2 = new ClassTypingContext(m1.getDeclaringType());
```

Comment:
Did you try to use existing `context`? I guess the existing one already contains all the necessary information, so there is no need to create new one for each method.

Reply:
done